### PR TITLE
Add solutions to exercises of Appendix A

### PR DESCRIPTION
**Pull request by Peter-VanWeert**

### CODE REVIEW

File: Exercises/NoModules/Appendix A/SolnA_04/SolnA_04.cpp [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  catch (const SmartException& exception)
+  {
+    std::cerr << "Something went wrong at line " << exception.where().line() << std::endl;
+  }
+
+  // Advanced use. This was no requirement of the exercise, 
+  // but it seems like something a smart exception could use...
+  try
+  {
+    SmartException exception{ "Throwing later...", {} };
```
Comment (MarcGregoire) [inline Exercises/NoModules/Appendix A/SolnA_04/SolnA_04.cpp:20]: technically, you don't have to specify the empty {} right.

File: Exercises/NoModules/Appendix A/SolnA_05/PrintThis.h [hunk 2]
```diff
@@ -0,0 +1,11 @@
+// Note: #include guards are strictly speaking optional for headers
+// that contain only function prototypes, 
+// but convention dictates you add the regardless.
```
Comment (MarcGregoire) [inline Exercises/NoModules/Appendix A/SolnA_05/PrintThis.h:3]: "them"

File: Exercises/NoModules/Appendix A/SolnA_05/PrintThat.h [hunk 3]
```diff
@@ -0,0 +1,11 @@
+// Note: #include guards are strictly speaking optional for headers
+// that contain only function prototypes, 
+// but convention dictates you add the regardless.
```
Comment (MarcGregoire) [inline Exercises/NoModules/Appendix A/SolnA_05/PrintThat.h:3]: "them"

File: Exercises/NoModules/Appendix A/SolnA_06/PrintThat.h [hunk 4]
```diff
@@ -0,0 +1,11 @@
+// Note: #include guards are strictly speaking optional for headers
+// that contain only function prototypes, 
+// but convention dictates you add the regardless.
```
Comment (MarcGregoire) [inline Exercises/NoModules/Appendix A/SolnA_06/PrintThat.h:3]: "them"

File: Exercises/NoModules/Appendix A/SolnA_06/PrintThis.h [hunk 5]
```diff
@@ -0,0 +1,11 @@
+// Note: #include guards are strictly speaking optional for headers
+// that contain only function prototypes, 
+// but convention dictates you add the regardless.
```
Comment (MarcGregoire) [inline Exercises/NoModules/Appendix A/SolnA_06/PrintThis.h:3]: "them"

File: Exercises/NoModules/Appendix A/SolnA_07A/PrintThis.h [hunk 6]
```diff
@@ -0,0 +1,11 @@
+// Note: #include guards are strictly speaking optional for headers
+// that contain only function prototypes, 
+// but convention dictates you add the regardless.
```
Comment (MarcGregoire) [inline Exercises/NoModules/Appendix A/SolnA_07A/PrintThis.h:3]: "them"

File: Exercises/NoModules/Appendix A/SolnA_07A/PrintThat.h [hunk 7]
```diff
@@ -0,0 +1,11 @@
+// Note: #include guards are strictly speaking optional for headers
+// that contain only function prototypes, 
+// but convention dictates you add the regardless.
```
Comment (MarcGregoire) [inline Exercises/NoModules/Appendix A/SolnA_07A/PrintThat.h:3]: "them"

File: Exercises/NoModules/Appendix A/SolnA_07B/PrintThis.h [hunk 8]
```diff
@@ -0,0 +1,11 @@
+// Note: #include guards are strictly speaking optional for headers
+// that contain only function prototypes, 
+// but convention dictates you add the regardless.
```
Comment (MarcGregoire) [inline Exercises/NoModules/Appendix A/SolnA_07B/PrintThis.h:3]: "them"

File: Exercises/NoModules/Appendix A/SolnA_07B/PrintThat.h [hunk 9]
```diff
@@ -0,0 +1,11 @@
+// Note: #include guards are strictly speaking optional for headers
+// that contain only function prototypes, 
+// but convention dictates you add the regardless.
```
Comment (MarcGregoire) [inline Exercises/NoModules/Appendix A/SolnA_07B/PrintThat.h:3]: "them"

File: Exercises/NoModules/Appendix A/SolnA_07B/Print.h [hunk 10]
```diff
@@ -0,0 +1,15 @@
+#ifndef PRINT_H
+#define PRINT_H
+
+#include <iostream>
+#include <string_view>
+
+inline int count = 0;
```
Comment (MarcGregoire) [inline Exercises/NoModules/Appendix A/SolnA_07B/Print.h:7]: uniform init

**Review by MarcGregoire — CHANGES_REQUESTED** [review]

File: Exercises/NoModules/Appendix A/SolnA_03/SmartException.cpp [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+SmartException::SmartException(const std::string& message, std::source_location location)
+  : std::logic_error{ message }
+  , m_location(location)
+{
+}
+
+void SmartException::throwFromHere(std::source_location location)
+{
+  m_location = std::move(location);
+  throw* this;
```
[the remark this answers is not in the document]
Comment (Peter-VanWeert) [reply Exercises/NoModules/Appendix A/SolnA_03/SmartException.cpp:12]: You mean the position of the *, right? Yup, very weird.

File: Exercises/NoModules/Appendix A/SolnA_03/SolnA_03.cpp [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+  catch (const SmartException& exception)
+  {
+    std::cerr << "Something went wrong at line " << exception.where().line() << std::endl;
+  }
+
+  // Advanced use. This was no requirement of the exercise, 
+  // but it seems like something a smart exception could use...
+  try
+  {
+    SmartException exception{ "Throwing later...", {} };
```
[the remark this answers is not in the document]
Comment (Peter-VanWeert) [reply Exercises/NoModules/Appendix A/SolnA_03/SolnA_03.cpp:21]: True. It's an optional performance optimisation. I commented on this in the member specification in the header. Added similar comment here as well.

File: Exercises/NoModules/Appendix A/SolnA_03/SmartException.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (MarcGregoire) [reply Exercises/NoModules/Appendix A/SolnA_03/SmartException.cpp]: yes, i would write "throw *this", not "throw* this"

**Pull request merged**